File: backend/src/StrideHR.Core/Models/ReportModels.cs
using StrideHR.Core.Enums;

namespace StrideHR.Core.Models;

public class ReportBuilderConfiguration
{
    public string DataSource { get; set; } = string.Empty;
    public List<ReportColumn> Columns { get; set; } = new();
    public List<ReportFilter> Filters { get; set; } = new();
    public List<ReportGrouping> Groupings { get; set; } = new();
    public List<ReportSorting> Sortings { get; set; } = new();
    public ReportChartConfiguration? ChartConfiguration { get; set; }
    public ReportPagination? Pagination { get; set; }
}

public class ReportColumn
{
    public string Name { get; set; } = string.Empty;
    public string DisplayName { get; set; } = string.Empty;
    public string DataType { get; set; } = string.Empty;
    public bool IsVisible { get; set; } = true;
    public int Order { get; set; }
    public string? Format { get; set; }
    public string? AggregateFunction { get; set; }
    public int? Width { get; set; }
    public string? Alignment { get; set; }
}

public class ReportFilter
{
    public string Column { get; set; } = string.Empty;
    public string Operator { get; set; } = string.Empty;
    public object? Value { get; set; }
    public string? LogicalOperator { get; set; } = "AND";
    public int Order { get; set; }
}

public class ReportGrouping
{
    public string Column { get; set; } = string.Empty;
    public string? AggregateFunction { get; set; }
    public int Order { get; set; }
}

public class ReportSorting
{
    public string Column { get; set; } = string.Empty;
    public string Direction { get; set; } = "ASC";
    public int Order { get; set; }
}

public class ReportChartConfiguration
{
    public ChartType Type { get; set; }
    public string Title { get; set; } = string.Empty;
    public string XAxisColumn { get; set; } = string.Empty;
    public string YAxisColumn { get; set; } = string.Empty;
    public string? SeriesColumn { get; set; }
    public Dictionary<string, object> Options { get; set; } = new();
    public List<string> Colors { get; set; } = new();
}

public class ReportPagination
{
    public int PageSize { get; set; } = 50;
    public bool EnablePaging { get; set; } = true;
}

public class ReportExecutionResult
{
    public bool Success { get; set; }
    public string? ErrorMessage { get; set; }
    public List<Dictionary<string, object>> Data { get; set; } = new();
    public int TotalRecords { get; set; }
    public TimeSpan ExecutionTime { get; set; }
    public Dictionary<string, object> Metadata { get; set; } = new();
}

public class ReportExportRequest
{
    public int ReportId { get; set; }
    public ReportExportFormat Format { get; set; }
    public Dictionary<string, object>? Parameters { get; set; }
    public string? FileName { get; set; }
    public bool IncludeCharts { get; set; } = true;
}

public class ReportScheduleRequest
{
    public int ReportId { get; set; }
    public string Name { get; set; } = string.Empty;
    public string CronExpression { get; set; } = string.Empty;
    public List<string> Recipients { get; set; } = new();
    public ReportExportFormat ExportFormat { get; set; }
    public string? EmailSubject { get; set; }
    public string? EmailBody { get; set; }
    public Dictionary<string, object>? Parameters { get; set; }
}

public class ReportDataSource
{
    public string Name { get; set; } = string.Empty;
    public string DisplayName { get; set; } = string.Empty;
    public string Description { get; set; } = string.Empty;
    public string ConnectionString { get; set; } = string.Empty;
    public string Query { get; set; } = string.Empty;
    public List<ReportDataSourceColumn> Columns { get; set; } = new();
}

public class ReportDataSourceColumn
{
    public string Name { get; set; } = string.Empty;
    public string DisplayName { get; set; } = string.Empty;
    public string DataType { get; set; } = string.Empty;
    public bool IsFilterable { get; set; } = true;
    public bool IsSortable { get; set; } = true;
    public bool IsGroupable { get; set; } = true;
    public List<string>? PossibleValues { get; set; }
}

File: frontend/src/app/models/report.models.ts
import { ReportType, ChartType, ReportExportFormat, ReportPermission } from '../enums/report.enums';

export interface ReportBuilderConfiguration {
  dataSource: string;
  columns: ReportColumn[];
  filters: ReportFilter[];
  groupings: ReportGrouping[];
  sortings: ReportSorting[];
  chartConfiguration?: ReportChartConfiguration;
  pagination?: ReportPagination;
}

export interface ReportColumn {
  name: string;
  displayName: string;
  dataType: string;
  isVisible: boolean;
  order: number;
  format?: string;
  aggregateFunction?: string;
  width?: number;
  alignment?: string;
}

export interface ReportFilter {
  column: string;
  operator: string;
  value: any;
  logicalOperator?: string;
  order: number;
}

export interface ReportGrouping {
  column: string;
  aggregateFunction?: string;
  order: number;
}

export interface ReportSorting {
  column: string;
  direction: string;
  order: number;
}

export interface ReportChartConfiguration {
  type: ChartType;
  title: string;
  xAxisColumn: string;
  yAxisColumn: string;
  seriesColumn?: string;
  options: { [key: string]: any };
  colors: string[];
}

export interface ReportPagination {
  pageSize: number;
  enablePaging: boolean;
}

export interface ReportExecutionResult {
  success: boolean;
  errorMessage?: string;
  data: { [key: string]: any }[];
  totalRecords: number;
  executionTime: number;
  metadata: { [key: string]: any };
}

export interface ReportDataSource {
  name: string;
  displayName: string;
  description: string;
  connectionString: string;
  query: string;
  columns: ReportDataSourceColumn[];
}

export interface ReportDataSourceColumn {
  name: string;
  displayName: string;
  dataType: string;
  isFilterable: boolean;
  isSortable: boolean;
  isGroupable: boolean;
  possibleValues?: string[];
}

export interface Report {
  id: number;
  name: string;
  description: string;
  type: ReportType;
  dataSource: string;
  configuration: string;
  filters: string;
  columns: string;
  chartConfiguration: string;
  isPublic: boolean;
  isScheduled: boolean;
  scheduleCron?: string;
  lastExecuted?: Date;
  status: string;
  createdBy: number;
  branchId?: number;
  createdAt: Date;
  updatedAt?: Date;
}

export interface CreateReportRequest {
  name: string;
  description: string;
  type: ReportType;
  configuration: ReportBuilderConfiguration;
}

export interface UpdateReportRequest {
  name: string;
  description: string;
  configuration: ReportBuilderConfiguration;
}

export interface ShareReportRequest {
  sharedWithUserId: number;
  permission: ReportPermission;
  expiresAt?: Date;
}

export interface ReportExportRequest {
  reportId: number;
  format: ReportExportFormat;
  parameters?: { [key: string]: any };
  fileName?: string;
  includeCharts: boolean;
}

export interface ReportScheduleRequest {
  reportId: number;
  name: string;
  cronExpression: string;
  recipients: string[];
  exportFormat: ReportExportFormat;
  emailSubject?: string;
  emailBody?: string;
  parameters?: { [key: string]: any };
}

export interface ReportSchedule {
  id: number;
  reportId: number;
  name: string;
  cronExpression: string;
  isActive: boolean;
  nextRunTime?: Date;
  lastRunTime?: Date;
  parameters?: string;
  recipients: string;
  exportFormat: ReportExportFormat;
  emailSubject?: string;
  emailBody?: string;
  createdBy: number;
  createdAt: Date;
  updatedAt?: Date;
}